File: go/poller_test.go
package mast

import (
	"context"
	"errors"
	"fmt"
	"testing"
	"time"
)

// Simulated seconds, so a ladder that spans ten minutes is tested in a
// millisecond and never depends on the machine being idle.
type fakeClock struct {
	at time.Time
}

func (c *fakeClock) now() time.Time { return c.at }

func (c *fakeClock) sleep(ctx context.Context, d time.Duration) error {
	c.at = c.at.Add(d)
	return nil
}

type stubChannel struct {
	clock  *fakeClock
	asks   []ask
	answer func(messageID string) (Status, error)
}

type ask struct {
	at        time.Time
	messageID string
}

func (s *stubChannel) Status(ctx context.Context, messageID string) (Status, error) {
	s.asks = append(s.asks, ask{at: s.clock.at, messageID: messageID})
	if s.answer == nil {
		return Status{State: "queued"}, nil
	}
	return s.answer(messageID)
}

func stubbed(t *testing.T) (*fakeClock, *stubChannel, *Poller) {
	t.Helper()
	clock := &fakeClock{at: time.Date(2026, 9, 17, 10, 0, 0, 0, time.UTC)}
	channel := &stubChannel{clock: clock}
	poller := &Poller{
		Budget:  PollBudgetPerMinute,
		channel: channel,
		open:    map[string]*watched{},
		last:    map[string]Status{},
		now:     clock.now,
		sleep:   clock.sleep,
	}
	return clock, channel, poller
}

// pollsOver ticks once per simulated second and reports the seconds that
// produced a poll.
func pollsOver(t *testing.T, clock *fakeClock, channel *stubChannel, poller *Poller, seconds int) []time.Time {
	t.Helper()
	var at []time.Time
	for i := 0; i < seconds; i++ {
		before := len(channel.asks)
		if _, err := poller.Tick(context.Background()); err != nil {
			t.Fatalf("Tick: %v", err)
		}
		if len(channel.asks) > before {
			at = append(at, clock.at)
		}
		clock.at = clock.at.Add(time.Second)
	}
	return at
}

func TestTheLadderSlowsDownAsThePageAges(t *testing.T) {
	cases := []struct {
		name    string
		seconds int
		after   time.Duration
		every   time.Duration
	}{
		{"three seconds while it is new", 30, 0, 3 * time.Second},
		{"ten seconds after half a minute", 120, 40 * time.Second, 10 * time.Second},
		{"half a minute after five", 500, 320 * time.Second, 30 * time.Second},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			clock, channel, poller := stubbed(t)
			poller.Budget = 1000
			started := clock.at
			if err := poller.Watch(messageID); err != nil {
				t.Fatal(err)
			}

			var late []time.Time
			for _, at := range pollsOver(t, clock, channel, poller, tc.seconds) {
				if at.Sub(started) > tc.after {
					late = append(late, at)
				}
			}
			if len(late) < 2 {
				t.Fatalf("only %d poll(s) in that band", len(late))
			}
			for i := 1; i < len(late); i++ {
				if gap := late[i].Sub(late[i-1]); gap != tc.every {
					t.Errorf("gap %d = %s, want %s", i, gap, tc.every)
				}
			}
		})
	}
}

func TestTwentyOpenPagesStillSpendHalfTheAllowance(t *testing.T) {
	clock, channel, poller := stubbed(t)
	for i := 0; i < 20; i++ {
		if err := poller.Watch(nthMessageID(i)); err != nil {
			t.Fatal(err)
		}
	}

	pollsOver(t, clock, channel, poller, 60)
	if len(channel.asks) > PollBudgetPerMinute {
		t.Errorf("spent %d requests in a minute, the whole key only gets %d", len(channel.asks), RateLimitPerMinute)
	}
}

func TestTheWindowRollsSoTheNextMinuteIsNotStarved(t *testing.T) {
	clock, channel, poller := stubbed(t)
	for i := 0; i < 20; i++ {
		if err := poller.Watch(nthMessageID(i)); err != nil {
			t.Fatal(err)
		}
	}

	pollsOver(t, clock, channel, poller, 121)
	if len(channel.asks) <= PollBudgetPerMinute {
		t.Errorf("spent %d requests in two minutes, so the window never reopened", len(channel.asks))
	}
	if len(channel.asks) > 3*PollBudgetPerMinute {
		t.Errorf("spent %d requests in two minutes", len(channel.asks))
	}
}

func TestTheOldestPageGetsTheLastRequest(t *testing.T) {
	clock, channel, poller := stubbed(t)
	poller.Budget = 1

	first, second := nthMessageID(1), nthMessageID(2)
	if err := poller.Watch(first); err != nil {
		t.Fatal(err)
	}
	clock.at = clock.at.Add(5 * time.Second)
	if err := poller.Watch(second); err != nil {
		t.Fatal(err)
	}

	if _, err := poller.Tick(context.Background()); err != nil {
		t.Fatal(err)
	}
	if len(channel.asks) != 1 || channel.asks[0].messageID != first {
		t.Errorf("asked about %v, want only the older page", channel.asks)
	}
}

func TestBothGetAskedAboutWhenThereIsRoom(t *testing.T) {
	clock, channel, poller := stubbed(t)

	first, second := nthMessageID(1), nthMessageID(2)
	if err := poller.Watch(first); err != nil {
		t.Fatal(err)
	}
	clock.at = clock.at.Add(5 * time.Second)
	if err := poller.Watch(second); err != nil {
		t.Fatal(err)
	}

	if _, err := poller.Tick(context.Background()); err != nil {
		t.Fatal(err)
	}
	if len(channel.asks) != 2 {
		t.Fatalf("asked about %d page(s), want 2", len(channel.asks))
	}
	if channel.asks[0].messageID != first {
		t.Errorf("asked about %s first, want the older page", channel.asks[0].messageID)
	}
}

func TestA429StopsTheWholeChannel(t *testing.T) {
	clock, channel, poller := stubbed(t)
	if err := poller.Watch(nthMessageID(1)); err != nil {
		t.Fatal(err)
	}
	if err := poller.Watch(nthMessageID(2)); err != nil {
		t.Fatal(err)
	}

	channel.answer = func(string) (Status, error) {
		return Status{}, &RateLimited{RetryAfter: 20 * time.Second}
	}
	if _, err := poller.Tick(context.Background()); err != nil {
		t.Fatal(err)
	}
	if len(channel.asks) != 1 {
		t.Fatalf("asked %d times, want to have stopped at the first 429", len(channel.asks))
	}

	channel.answer = nil
	clock.at = clock.at.Add(5 * time.Second)
	if _, err := poller.Tick(context.Background()); err != nil {
		t.Fatal(err)
	}
	if len(channel.asks) != 1 {
		t.Fatalf("asked %d times, want the channel still paused", len(channel.asks))
	}

	clock.at = clock.at.Add(20 * time.Second)
	if _, err := poller.Tick(context.Background()); err != nil {
		t.Fatal(err)
	}
	if len(channel.asks) != 3 {
		t.Fatalf("asked %d times, want both pages once the pause is over", len(channel.asks))
	}
}

func TestABlipLeavesThePageOpen(t *testing.T) {
	clock, channel, poller := stubbed(t)
	if err := poller.Watch(messageID); err != nil {
		t.Fatal(err)
	}

	channel.answer = func(string) (Status, error) {
		return Status{}, fmt.Errorf("%w: connection reset", ErrUnreachable)
	}
	if _, err := poller.Tick(context.Background()); err != nil {
		t.Fatal(err)
	}
	if poller.Open() != 1 {
		t.Fatal("one failed poll closed the page")
	}

	channel.answer = func(string) (Status, error) { return Status{State: "acked"}, nil }
	clock.at = clock.at.Add(5 * time.Second)
	if _, err := poller.Tick(context.Background()); err != nil {
		t.Fatal(err)
	}
	if poller.Open() != 0 {
		t.Error("the ack did not settle the page")
	}
}

func TestAMessageTheChannelDisownsIsDropped(t *testing.T) {
	_, channel, poller := stubbed(t)
	if err := poller.Watch(messageID); err != nil {
		t.Fatal(err)
	}

	channel.answer = func(string) (Status, error) {
		return Status{}, fmt.Errorf("%w: No such message.", ErrUnknownMessage)
	}
	if _, err := poller.Tick(context.Background()); err != nil {
		t.Fatal(err)
	}
	if poller.Open() != 0 {
		t.Error("kept polling a message the channel does not have")
	}
}

func TestAnAckTakesThePageOffTheList(t *testing.T) {
	_, channel, poller := stubbed(t)
	if err := poller.Watch(messageID); err != nil {
		t.Fatal(err)
	}

	channel.answer = func(string) (Status, error) {
		return Status{
			State:      "acked",
			ReceivedAt: time.Date(2026, 9, 17, 10, 0, 0, 0, time.UTC),
			AckedAt:    time.Date(2026, 9, 17, 10, 0, 12, 0, time.UTC),
			AckedBy:    "denis",
		}, nil
	}

	settled, err := poller.Tick(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	if poller.Open() != 0 {
		t.Error("the page is still being watched")
	}
	status, ok := settled[messageID]
	if !ok {
		t.Fatalf("settled = %v", settled)
	}
	if open, _ := status.OpenFor(); open != 12*time.Second || status.AckedBy != "denis" {
		t.Errorf("settled with %+v", status)
	}
}

func TestRunReturnsOnceNothingIsOpen(t *testing.T) {
	_, channel, poller := stubbed(t)
	if err := poller.Watch(messageID); err != nil {
		t.Fatal(err)
	}
	channel.answer = func(string) (Status, error) { return Status{State: "resolved"}, nil }

	settled, err := poller.Run(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	if len(settled) != 1 {
		t.Fatalf("settled = %v", settled)
	}
}

func TestACancelledRunLeavesThePageWatched(t *testing.T) {
	_, _, poller := stubbed(t)
	if err := poller.Watch(messageID); err != nil {
		t.Fatal(err)
	}

	ctx, cancel := context.WithCancel(context.Background())
	poller.sleep = func(context.Context, time.Duration) error {
		cancel()
		return ctx.Err()
	}

	settled, err := poller.Run(ctx)
	if !errors.Is(err, context.Canceled) {
		t.Fatalf("err = %v, want the cancellation", err)
	}
	if len(settled) != 0 || poller.Open() != 1 {
		t.Errorf("settled = %v, open = %d", settled, poller.Open())
	}
	if status, ok := poller.Last(messageID); !ok || status.State != "queued" {
		t.Errorf("Last() = %+v, %v", status, ok)
	}
}

func TestWatchingTheSamePageTwiceWatchesOnePage(t *testing.T) {
	_, _, poller := stubbed(t)
	for i := 0; i < 2; i++ {
		if err := poller.Watch(messageID); err != nil {
			t.Fatal(err)
		}
	}
	if poller.Open() != 1 {
		t.Errorf("open = %d, want 1", poller.Open())
	}
}

func TestWatchRefusesSomethingThatIsNotAMessageID(t *testing.T) {
	_, _, poller := stubbed(t)
	if err := poller.Watch("mm_nope"); err == nil {
		t.Fatal("a bad message id was accepted")
	}
}

func nthMessageID(n int) string {
	return fmt.Sprintf("mm_%032x", n)
}

File: javascript/test/poller.test.js
import assert from "node:assert/strict";
import { describe, test } from "node:test";

import {
  POLL_BUDGET_PER_MIN,
  Poller,
  RateLimited,
  Rejected,
  Status,
  UnknownMessage,
} from "../src/mast.js";

const FIRST = "mm_" + "7a3e".repeat(8);
const SECOND = "mm_" + "91bc".repeat(8);

// Simulated seconds, so a ladder that spans ten minutes is tested in a
// millisecond and never depends on the machine being idle.
class Clock {
  constructor() {
    this.t = 1000;
  }

  now = () => this.t;

  sleep = async (ms) => {
    this.t += ms / 1000;
  };
}

class StubChannel {
  constructor(clock) {
    this.clock = clock;
    this.asks = [];
    this.answer = () => new Status({ state: "queued" });
  }

  async status(messageId) {
    this.asks.push({ at: this.clock.t, messageId });
    return this.answer(messageId);
  }
}

function setup(options = {}) {
  const clock = new Clock();
  const channel = new StubChannel(clock);
  const poller = new Poller(channel, { now: clock.now, sleep: clock.sleep, ...options });
  return { clock, channel, poller };
}

// Tick once per simulated second and report the seconds that produced a poll.
async function pollsOver(clock, channel, poller, seconds) {
  const at = [];
  for (let i = 0; i < seconds; i += 1) {
    const before = channel.asks.length;
    await poller.tick();
    if (channel.asks.length > before) at.push(clock.t);
    clock.t += 1;
  }
  return at;
}

describe("the ladder", () => {
  test("asks every three seconds while the page is new", async () => {
    const { clock, channel, poller } = setup();
    poller.watch(FIRST);
    const at = await pollsOver(clock, channel, poller, 30);
    assert.deepEqual(gaps(at), Array(at.length - 1).fill(3));
  });

  test("drops to ten seconds after half a minute", async () => {
    const { clock, channel, poller } = setup();
    poller.watch(FIRST);
    const at = await pollsOver(clock, channel, poller, 120);
    const late = at.filter((t) => t - 1000 > 40);
    assert.deepEqual(gaps(late), Array(late.length - 1).fill(10));
  });

  test("drops to half a minute once the page is five minutes old", async () => {
    const { clock, channel, poller } = setup({ budgetPerMinute: 1000 });
    poller.watch(FIRST);
    const at = await pollsOver(clock, channel, poller, 500);
    const late = at.filter((t) => t - 1000 > 320);
    assert.deepEqual(gaps(late), Array(late.length - 1).fill(30));
  });
});

describe("the budget", () => {
  test("twenty open pages still spend at most half the key's allowance", async () => {
    const { clock, channel, poller } = setup();
    for (let i = 0; i < 20; i += 1) poller.watch(messageId(i));
    await pollsOver(clock, channel, poller, 60);
    assert.ok(
      channel.asks.length <= POLL_BUDGET_PER_MIN,
      `spent ${channel.asks.length} requests in a minute`,
    );
  });

  test("the window rolls, so the next minute is not starved", async () => {
    const { clock, channel, poller } = setup();
    for (let i = 0; i < 20; i += 1) poller.watch(messageId(i));
    await pollsOver(clock, channel, poller, 121);
    assert.ok(channel.asks.length > POLL_BUDGET_PER_MIN);
    assert.ok(channel.asks.length <= POLL_BUDGET_PER_MIN * 3);
  });

  test("the oldest page gets the last request", async () => {
    const { clock, channel, poller } = setup({ budgetPerMinute: 1 });
    poller.watch(FIRST);
    clock.t += 5;
    poller.watch(SECOND);
    await poller.tick();
    assert.deepEqual(channel.asks.map((a) => a.messageId), [FIRST]);
  });

  test("both get asked about when there is room for both", async () => {
    const { clock, channel, poller } = setup();
    poller.watch(FIRST);
    clock.t += 5;
    poller.watch(SECOND);
    await poller.tick();
    assert.deepEqual(channel.asks.map((a) => a.messageId), [FIRST, SECOND]);
  });
});

describe("backing off", () => {
  test("a 429 stops the whole channel, not the one page", async () => {
    const { clock, channel, poller } = setup();
    poller.watch(FIRST);
    poller.watch(SECOND);
    channel.answer = () => {
      throw new RateLimited(20);
    };
    await poller.tick();
    assert.equal(channel.asks.length, 1);

    channel.answer = () => new Status({ state: "queued" });
    clock.t += 5;
    await poller.tick();
    assert.equal(channel.asks.length, 1);

    clock.t += 20;
    await poller.tick();
    assert.equal(channel.asks.length, 3);
  });

  test("a blip leaves the page open", async () => {
    const { clock, channel, poller } = setup();
    poller.watch(FIRST);
    channel.answer = () => {
      throw new Error("connection reset");
    };
    await poller.tick();
    assert.equal(poller.openCount, 1);

    channel.answer = () => new Status({ state: "acked" });
    clock.t += 5;
    await poller.tick();
    assert.equal(poller.openCount, 0);
  });

  test("a message the channel disowns is dropped", async () => {
    const { channel, poller } = setup();
    poller.watch(FIRST);
    channel.answer = () => {
      throw new UnknownMessage("No such message.");
    };
    await poller.tick();
    assert.equal(poller.openCount, 0);
  });
});

describe("settling", () => {
  test("an ack takes the page off the list", async () => {
    const { channel, poller } = setup();
    poller.watch(FIRST);
    channel.answer = () =>
      new Status({
        state: "acked",
        received_at: "2026-09-17T10:00:00Z",
        acked_at: "2026-09-17T10:00:12Z",
        acked_by: "denis",
      });
    const settled = await poller.tick();
    assert.equal(poller.openCount, 0);
    assert.equal(settled.get(FIRST).ackedBy, "denis");
    assert.equal(settled.get(FIRST).openFor, 12);
  });

  test("run returns once nothing is open", async () => {
    const { channel, poller } = setup();
    poller.watch(FIRST);
    channel.answer = () => new Status({ state: "resolved" });
    const settled = await poller.run();
    assert.deepEqual([...settled.keys()], [FIRST]);
  });

  test("a timeout leaves the page open rather than failing it", async () => {
    const { poller } = setup();
    poller.watch(FIRST);
    const settled = await poller.run({ timeout: 5000 });
    assert.equal(settled.size, 0);
    assert.equal(poller.openCount, 1);
    assert.equal(poller.lastStatus(FIRST).state, "queued");
  });

  test("watching the same page twice watches one page", () => {
    const { poller } = setup();
    poller.watch(FIRST);
    poller.watch(FIRST);
    assert.equal(poller.openCount, 1);
  });

  test("it will not watch something that is not a message id", () => {
    const { poller } = setup();
    assert.throws(() => poller.watch("mm_nope"), Rejected);
  });
});

function gaps(values) {
  return values.slice(1).map((value, i) => value - values[i]);
}

function messageId(n) {
  return "mm_" + String(n).padStart(4, "0").repeat(8);
}
